lib/resque/data_store.rb

```diff
@@ -122,11 +122,45 @@ def push_to_queue(queue,encoded_item)
         end
       end
 
-      # Pop whatever is on queue
-      def pop_from_queue(queue)
-        @redis.lpop(redis_key_for_queue(queue))
+      # MLPOP key [key ..]
+      #
+      # LPOP the first available element by probing each key
+      # in the specified order.
+      #
+      # Moving the scheduling logic to the server reduces
+      # the required round-trips especially for workers with
+      # many queues in their watchlists.
+      #
+      # Returns a array of [key-index, element value] or nil
+      # if no element was found.
+      #
+      # Note that we choose to return the key index instead
+      # of the key itself: Redis::Namespace converts the keys
+      # passed to the script (adding a prefix), and we would
+      # have to reverse map the return values. We avoid that
+      # by returning the index instead.
+      def mlpop(*queues, keys:)
+        @redis.evalsha(mlpop_sha, keys: keys)
+      rescue Redis::CommandError => e
+        # If the script  is not found (e.g. when the server
+        # is restarted) reload it.
+        if e.message.include?("NOSCRIPT")
+          mlpop_sha!
+          retry
+        end
+
+        raise e
       end
 
+      def pop_from_queue(*queues)
```

Comment:
[nit] We could use the plural here for `queues` and implement the `pop_from_queue` calling this one to avoid breaking the api.

```suggestion
      def pop_from_queues(*queues)
        ...
      end
      
      def pop_from_queue(queue)
        pop_from_queues(queue)
      end
```